tests/scattering1d/test_jtfs.py

```diff
@@ -0,0 +1,372 @@
+    T0F1 = coeff_energy(outs[(Ts[0], Fs[1])], metas[(Ts[0], Fs[1])])
+    T1F0 = coeff_energy(outs[(Ts[1], Fs[0])], metas[(Ts[1], Fs[0])])
+    T1F1 = coeff_energy(outs[(Ts[1], Fs[1])], metas[(Ts[1], Fs[1])])
+
+    if metric_verbose:
+        print("\nGlobal averaging reldiffs:")
+
+    th = .15
+    for pair in T0F0:
+        ref = T0F0[pair]
+        reldiff01 = abs(T0F1[pair] - ref) / ref
+        reldiff10 = abs(T1F0[pair] - ref) / ref
+        reldiff11 = abs(T1F1[pair] - ref) / ref
+        assert reldiff01 < th, "%s > %s | %s" % (reldiff01, th, pair)
+        assert reldiff10 < th, "%s > %s | %s" % (reldiff10, th, pair)
+        assert reldiff11 < th, "%s > %s | %s" % (reldiff11, th, pair)
+
+        if metric_verbose:
+            print("(01, 10, 11) = ({:.2e}, {:.2e}, {:.2e}) | {}".format(
+                reldiff01, reldiff10, reldiff11, pair))
+
+
+def test_implementation():
+    """Test that every `implementation` kwarg works."""
+    N = 512
+    x = echirp(N)
+
+    for implementation in range(1, 6):
+        jtfs = TimeFrequencyScattering1D(shape=N, J=4, Q=2,
+                                         implementation=implementation,
+                                         frontend=default_backend)
+        _ = jtfs(x)
+
+
+def test_pad_mode_fr():
+    """Test that functional `pad_mode_fr` works."""
+    from kymatio.scattering1d.core.timefrequency_scattering1d import _right_pad
+    N = 512
+    x = echirp(N)
+
+    kw = dict(shape=N, J=4, Q=2, frontend=default_backend, out_type='array',
+              max_pad_factor_fr=1)
+    jtfs0 = TimeFrequencyScattering1D(**kw, pad_mode_fr='zero')
+    jtfs1 = TimeFrequencyScattering1D(**kw, pad_mode_fr=_right_pad)
+
+    out0 = jtfs0(x)
+    out1 = jtfs1(x)
+    assert np.allclose(out0, out1)
+
+
+def test_no_second_order_filters():
+    """Reproduce edge case: configuration yields no second-order wavelets
+    so can't do JTFS.
+    """
+    with pytest.raises(ValueError) as record:
+        _ = TimeFrequencyScattering1D(shape=8192, J=1, Q=2, r_psi=.9,
+                                      frontend=default_backend)
+    assert "no second-order filters" in record.value.args[0]
+
+
+def concat_joint(Scx):
+    Scx = drop_batch_dim_jtfs(Scx)
+    k = list(Scx)[0]
+    out_type = ('list' if (isinstance(Scx[k], list) and 'coef' in Scx[k][0]) else
+                'array')
+    if out_type == 'array':
+        return np.vstack([c for pair, c in Scx.items()
+                          if pair not in ('S0', 'S1')])
+    return np.vstack([c['coef'] for pair, coeffs in Scx.items()
+                      for c in coeffs if pair not in ('S0', 'S1')])
+
+
+def assert_pad_difference(jtfs, test_params_str):
+    assert not all(
+        J_pad_fr == jtfs.J_pad_frs_max
+        for J_pad_fr in jtfs.J_pad_frs if J_pad_fr != -1
+        ), "\n{}\nJ_pad_fr={}\nN_frs={}".format(
+            test_params_str, jtfs.J_pad_frs, jtfs.N_frs)
+
+
+if __name__ == '__main__':
+    if run_without_pytest:
```

Comment:
what if you accidentally omit a test in this block? that means no coverage

This does not have to be commited since it's a developer feature. 

It is not much effort to add for your personal convenience.

Reply:
> what if you accidentally omit a test in this block?

`run_without_pytest == False` when committed, always. This can be refined further to override in case someone commits `True`, by [detecting](https://stackoverflow.com/q/25188119/10133797) the runtime environment, or by setting one global value in `tests/__init__.py`, which I'm willing to do - but saw no need for just two files.

Also Github notified me of none of this, helps to @ @rastegah